src/CMakeLists.txt

```diff
@@ -78,214 +89,309 @@ option(TPL_LAPACK_LIBRARIES          "Optional list of absolute paths to LAPACK
-endif()
+endif ()
 
 if (HYPRE_ENABLE_HYPRE_BLAS)
   set(HYPRE_USING_HYPRE_BLAS ON CACHE BOOL "" FORCE)
-endif()
+endif ()
 
 if (HYPRE_ENABLE_HYPRE_LAPACK)
   set(HYPRE_USING_HYPRE_LAPACK ON CACHE BOOL "" FORCE)
-endif()
+endif ()
 
 if (HYPRE_ENABLE_PERSISTENT_COMM)
   set(HYPRE_USING_PERSISTENT_COMM ON CACHE BOOL "" FORCE)
-endif()
+endif ()
 
 if (HYPRE_WITH_MPI)
+  set(HYPRE_HAVE_MPI ON CACHE BOOL "" FORCE)
   set(HYPRE_SEQUENTIAL OFF CACHE BOOL "" FORCE)
-else()
+else ()
   set(HYPRE_SEQUENTIAL ON CACHE BOOL "" FORCE)
-endif()
+endif ()
 
 if (HYPRE_WITH_OPENMP)
-  set (HYPRE_USING_OPENMP ON CACHE BOOL "" FORCE)
-endif()
+  set(HYPRE_USING_OPENMP ON CACHE BOOL "" FORCE)
+endif ()
 
 if (HYPRE_WITH_HOPSCOTCH)
-  set (HYPRE_HOPSCOTCH ON CACHE BOOL "" FORCE)
-endif()
+  set(HYPRE_HOPSCOTCH ON CACHE BOOL "" FORCE)
+endif ()
 
 if (HYPRE_WITH_DSUPERLU)
-  set (HYPRE_USING_DSUPERLU ON CACHE BOOL "" FORCE)
-  set (HYPRE_USING_HYPRE_BLAS OFF CACHE BOOL "" FORCE)
-  set (HYPRE_USING_HYPRE_LAPACK OFF CACHE BOOL "" FORCE)
-endif()
+  set(HYPRE_USING_DSUPERLU ON CACHE BOOL "" FORCE)
+  set(HYPRE_USING_HYPRE_BLAS OFF CACHE BOOL "" FORCE)
+  set(HYPRE_USING_HYPRE_LAPACK OFF CACHE BOOL "" FORCE)
+endif ()
 
 if (HYPRE_ENABLE_FEI)
-  set (HYPRE_USING_FEI ON CACHE BOOL "" FORCE)
+  set(HYPRE_USING_FEI ON CACHE BOOL "" FORCE)
   message(WARNING "CMake support for FEI is not complete!")
-endif()
+endif ()
 
 if (HYPRE_WITH_CALIPER)
-  set (HYPRE_USING_CALIPER ON CACHE BOOL "" FORCE)
-endif()
+  set(HYPRE_USING_CALIPER ON CACHE BOOL "" FORCE)
+endif ()
 
 if (HYPRE_SHARED OR HYPRE_BIGINT OR HYPRE_SINGLE OR HYPRE_LONG_DOUBLE)
   # FEI doesn't currently compile with shared
-  set (HYPRE_USING_FEI OFF CACHE BOOL "" FORCE)
+  set(HYPRE_USING_FEI OFF CACHE BOOL "" FORCE)
 endif ()
 
 if (HYPRE_SEQUENTIAL)
-  set (HYPRE_BUILD_EXAMPLES OFF CACHE BOOL "" FORCE)
+  set(HYPRE_BUILD_EXAMPLES OFF CACHE BOOL "" FORCE)
 endif ()
 
-add_library (HYPRE)
+# CUDA
+if (HYPRE_WITH_CUDA)
+  enable_language(CXX)
+  message(STATUS "Enabled support for CXX.")
+
+  # Enforce C++11
+  if (NOT CMAKE_CXX_STANDARD OR CMAKE_CXX_STANDARD LESS 11)
+    set(CMAKE_CXX_STANDARD 11)
+  endif ()
+  set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+  message(STATUS "Using CXX standard: c++${CMAKE_CXX_STANDARD}")
+
+  # Use ${CMAKE_CXX_COMPILER} as the cuda host compiler.
+  if (NOT CMAKE_CUDA_HOST_COMPILER)
+    set(CMAKE_CUDA_HOST_COMPILER ${CMAKE_CXX_COMPILER})
+  endif ()
+
+  # Check if CUDA is available, then enable it
+  include(CheckLanguage)
+  check_language(CUDA)
+  if (CMAKE_CUDA_COMPILER)
+
+    enable_language(CUDA)
+    message(STATUS "Enabled support for CUDA.")
+
+    if (NOT CMAKE_CUDA_STANDARD OR CMAKE_CUDA_STANDARD EQUAL 98)
+      set(CMAKE_CUDA_STANDARD 11)
+    endif ()
+
+    set(CMAKE_CUDA_STANDARD_REQUIRED TRUE)
+
+    set(HYPRE_USING_CUDA ON CACHE BOOL "" FORCE)
+    set(HYPRE_USING_GPU ON CACHE BOOL "" FORCE)
+
+    if (HYPRE_ENABLE_UNIFIED_MEMORY)
```

Comment:
@liruipeng Is this the correct logic for the `*_MEMORY` variable?

Reply:
This looks good .